[2단계 - 자동차 경주 리팩터링] 렉스(오성원) 미션 제출합니다.

Seongwon97 opened the pull request:
안녕하세요 미르😁

1단계 피드백 때 반영하지 못하였던 랜덤값 생성을 하는 goForward 메서드 테스트(https://github.com/woowacourse/java-racingcar/pull/326#discussion_r805077065) 와 객체의 정보를 가져오는 것이 아닌 메시지를 보내는 방식으로의 변환(https://github.com/woowacourse/java-racingcar/pull/326#discussion_r805075064) 을 드디어 이해하고 반영하게 되었습니다.

또한 여러 내용을 학습하며 RandomNumber 인스턴스의 불필요한 중복 생성을 막아보고자 싱글톤 패턴을 적용하였으며 Car리스트를 일급 컬렉션으로 분리해보는 리펙터링도 진행해봤습니다.

### 2단계를 진행하며 궁금한 점을 정리해봤습니다.
- 리펙토링을 진행하며 RacingGame.java파일에 위치하였던 List<Car> 변수를 Cars.java라는 일급 컬렉션으로 분리하고 해당 클래스 내에서 List<Car>의 정보로 알아낼 수 있는 `턴 별 결과`, `우승자 찾기`를 비롯해 여러 RacingGame.java파일에 위치하였던 여러 메서드를 이동시킨 후 추가적인 리펙터링을 진행했습니다. 
여기서 궁금한 점은 제가 현재 models에 위치한 Cars.java에서 턴 별 결과와 우승자를 찾은 결과를 String으로 return하고 해당 결과를 Output.java의 메서드에서 매개변수로 받아 출력을 하는 방식으로 변경을 하게 되었는데, 이러한 경우도 view와 model이 엮여있는 형태로 봐야하는지 궁금합니다🧐

미션을 진행하며 궁금한 점들이 매우 많았는데..이렇게 막상 PR을 보내려니 기억이 잘 안나네요😅
미리 메모를 해뒀어야했는데...슬프네요🥲

감사합니다 🙇🏻‍♂️

Seongwon97 committed:
refactor(RacingGame.java, Car.java): 필드들의 접근 지정자, 불변 설정

- Car.java파일의 인스턴스 변수들의 접근 지정자를 private로 변경
- RacingGame.java파일의 List<Car>에 final을 붙여 불변으로 변경

Seongwon97 committed:
docs(README.md): 난수 조건 수정

- 난수가 생성되는 조건에 대한 내용 수정

Seongwon97 committed:
refactor(Output.java): 출력문의 줄바꿈을 lineSeparator()를 사용하도록 변경

- OS마다 개행이 달라질 수 있어 줄바꿈인 "\n"을 System.lineSeparator()로 변경

Seongwon97 committed:
refactor(/src): import순서 조정, 불변 값들에 final추가

- 불필요한 import문 제거 및 import들 순서 조정
- 불변 값들에 final 예약어를 추가

Seongwon97 committed:
refactor(InputValidation.java): 검증 메서드 이름의 접두사 통일

- 메서드 이름을 10~20자리 사이로 축약
- 해당 클래스 내의 메서드 이름의 접두사를 check로 통일

Seongwon97 committed:
refactor(RandomNumber.java): RandomNumber에 싱글톤 패턴 적용

- RandomNumber에 싱글톤 패턴을 적용하여 Car객체마다 생성되던 인스턴스가
프로그램 전체에서 하나만 생성되어 공유되도록 변경

Seongwon97 committed:
refactor(RandomNumber.java): 인스턴스를 정적 변수에서 선언하도록 변경

- 멀티 쓰레드 환경에서의 동시성 문제를 막기 위해 인스턴스를 정적 변수에서 선언하도록 변경

Seongwon97 committed:
refactor(Car.java): toString재정의, ui와 domain분리

- 재정의한 toString을 로직에 사용하며 연결되었던 ui와 domain을 분리
- name정보만 출력하던 toString()재정의를 디버깅, 로깅을 할 때 객체의 정보를
쉽게 파악하기 위해 name, position정보를 둘 다 출력하도록 수정

Seongwon97 committed:
refactor(Car.java): isWinner메서드를 통해 데이터를 가지는 객체가 일하도록 변경

- isWinner()메서드를 생성하여 car객체가 가장 멀리간 위치 정보를 받고
우승자인지 판단하여 결과를 retuen하도록 변경
- RacingGame.java의 findFarestPosition메서드에서 sorting된 자동차 객체를 저장하는
list 인스턴스 필드를 Car이 아닌 위치값만 저장하도록 하여 메모리 소비 감소

Seongwon97 committed:
refactor(Car.java): 난수 관련 부분을 메서드에서 분리

- goForward()메서드의 난수 생성 관련 부분을 분리하여
테스트 가능한 부분에 대해서 단위테스트가 가능하도록 변경

Seongwon97 committed:
refactor(Cars.java, RacingGame.java): car리스트를 일급 컬렉션으로 분리

- car리스트를 일급 컬렉션으로 분리

Seongwon97 committed:
refactor(Cars.java): 턴 결과를 String리스트로 반환하도록 변경

- Cars.java객체에서 turnResult를 통해 턴 결과를 String리스트로 반환하고
Output.java에서는 해당 String리스트를 출력하도록 변환

Seongwon97 committed:
refactor(/racingcar): indent통일 및 불필요한 import제거

Seongwon97 committed:
fix(RandomNumberTest.java): 잘못된 난수 범위 수정

- 1~9까지의 범위였던 난수 범위를 0~9로 수정

Seongwon97 committed:
test(CarTest.java): 자동차가 우승자가 맞는지 확인하는 테스트 추가

- Car.java의 isWinner메서드 테스트

Seongwon97 committed:
test(CarsTest.java): Cars.java의 메서드 테스트 추가

- findWinners()메서드가 우승자를 옳바르게 찾는지 테스트
- turnResult()메서드가 현재 턴의 결과를 옳바르게 반환하는지 테스트 추가

Seongwon97 committed:
refactor(CarsTest.java): 불필요한 import제거

Seongwon97 committed:
refactor(src/): 모든 파일의 공백 라인 통일

- 클래스 시작 시 첫 줄 공백
- 클래스 닫히는 괄호 이전에는 공백 라인 X

ddu0422 commented on src/main/java/racingcar/models/Car.java:
```diff
@@ -1,53 +1,38 @@
+        if (randomNumber >= FORWARD_STANDARD) {
+            position++;
+        }
+    }
```
👍🏻 테스트 하기 편한 구조로 바뀌었군요!
고민을 많이 하셨군요 ㅎ_ㅎ.. 힘드셨을텐데 고생하셨어요 :)

추가로 드리는 질문입니다!
goForward라는 메서드의 파라미터 변수명이 randomNumber인데 해당 메서드에서 random인지 아닌지 알 필요가 있을까요 🤔

Seongwon97 replied:
오늘 네오의 강의를 들으며 확실히 이해가 된 것 같습니다!!

한 줄 정리 : 모든 변수와 메서드 명은 해당 클래스의 입장?기준?에서 이름을 정하는 것이 좋다❗️

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -1,27 +1,23 @@
         this.output = output;
-        cars = new ArrayList<>();
         repeats = 0;
     }
```
RacingGame이 Input과 Output을 가지는 구조는 현재 다음과 같아요!

![스크린샷 2022-02-17 00 40 13](https://user-images.githubusercontent.com/43198414/154494127-f66d522a-cd8e-43c1-9359-577d86ca5d02.png)

이런 식으로 변경해보면 어떨까요?? Input으로 받은 값을 넣어보는거죠!

![스크린샷 2022-02-17 00 40 19](https://user-images.githubusercontent.com/43198414/154494184-ca96b476-a45e-48b0-b242-3528a26e0f31.png)

Seongwon97 replied:
미르의 그림 덕분에 드디어 이해가 됐습니다..😅

Controller와 View의 관계를 분리하기 위해 Input, Output 내부의 메서드들을 정적 메서드로 변경하여 Controller에서는 해당 메서드를 정적으로 호출하여 사용하도록 변경해봤습니다!!

ddu0422 commented on src/main/java/racingcar/models/Car.java:
```diff
@@ -1,53 +1,38 @@
+
+    public boolean isWinner(final int farthestPosition) {
+        return position == farthestPosition;
+    }
```
메시지를 전달하셨군요 !__! 👍🏻 👍🏻
하지만 메서드 명과 매개변수 명을 조금 고민해보셨으면 좋겠어요 :)

과연.. 자동차 객체가 Winner임을 알 필요가 있을까요? 가장 멀리 있는 위치임을 알 필요가 있을까요 ? 🤔

Seongwon97 replied:
해당 부분도 Car의 입장에서 생각을 하여 아래와 같이 위치값을 받아 같은 위치인지 판단!하는 이름의 메서드로 변경해보았습니다!!

```java
    public boolean isSamePosition(final int position) {
        return this.position == position;
    }
```

ddu0422 commented on src/main/java/racingcar/models/Cars.java:
```diff
@@ -0,0 +1,68 @@
+    public Cars(List<Car> cars) {
+        this.randomNumber = RandomNumber.getInstance();
+        this.cars = cars;
+    }
```
일급 컬렉션을 작성해주셨군요 !_!
하지만 해당 객체가 RandomNumber와 강하게 결합을 하고 있네요 😭
만약 다른 RandomNumber를 사용하고 싶다면 변경하기 어렵지 않을까요 ?_?

ddu0422 commented on src/main/java/racingcar/models/Cars.java:
```diff
@@ -0,0 +1,68 @@
+                .map(Car::getPosition)
+                .collect(Collectors.toList());
+
+        if (!sortedCars.isEmpty()) {
+            return sortedCars.get(0);
+        }
+        return 0;
```
💯 💯

ddu0422 commented on src/main/java/racingcar/models/Cars.java:
```diff
@@ -0,0 +1,68 @@
+            result.append(CAR_POSITION);
+        }
+        return result.toString();
+    }
```
만약 출력(Output)이 바뀌게 되면 해당 부분이 변경되겠군요!
출력이 바뀌는 거라면 Output(UI)이 바뀌어야지 Domain이 변경되어도 괜찮을까요?_?

Seongwon97 replied:
그렇군요..출력과 연관된 내용 중 하나라도 도메인에 들어가게 된다면 그 둘은 분리된게 아니고 연관되게 되는거 같네요.

미르의 말을 듣고보니 carTurnReut메서드 뿐만 아니라 findWinner메서드도 출력 String을 도메인에서 가공하고 return하여 view와 너무 깊게 연결되어 있던 것 같습니다.

도메인에서는 관련 값들을 찾아서 List로 return하고 controller에서 Output메서드를 호출하여 해당 결과 String가공은 Output.java파일에서 하도록 수정해봤습니다🔥

ddu0422 reviewed (changes requested):
안녕하세요 렉스 :) 미션 진행하시느라 고생하셨어요!
UI와 분리하는 방법을 좀 더 고민해보시면 좋을 것 같아요 😄
추가적으로 인스턴스 변수를 보며 접근제한자가 package-private인 부분을 적절한 수준으로 변경해보는 건 어떨까요?

## Q&A
> 리펙토링을 진행하며 RacingGame.java파일에 위치하였던 List 변수를 Cars.java라는 일급 컬렉션으로 분리하고 해당 클래스 내에서 List의 정보로 알아낼 수 있는 턴 별 결과, 우승자 찾기를 비롯해 여러 RacingGame.java파일에 위치하였던 여러 메서드를 이동시킨 후 추가적인 리펙터링을 진행했습니다.
여기서 궁금한 점은 제가 현재 models에 위치한 Cars.java에서 턴 별 결과와 우승자를 찾은 결과를 String으로 return하고 해당 결과를 Output.java의 메서드에서 매개변수로 받아 출력을 하는 방식으로 변경을 하게 되었는데, 이러한 경우도 view와 model이 엮여있는 형태로 봐야하는지 궁금합니다🧐

조금 쉽게 표현하려다보니 view와 model이 엮여있다가 표현이 부적절했던 것 같군요 ㅎ_ㅎ..
domain(model)은 Input / Output과 관련이 없도록 하고 Output은 Domain을 알고 있어도 괜찮을 것 같아요 :)
해당 부분에 추가로 코멘트 드렸어요.

궁금하거나 모호한 부분이 있다면 언제든디 코멘트나 DM 주세요 !_!
마지막까지 화이팅해보아요 🔥

Seongwon97 committed:
refactor(Car.java): 메서드와 변수명 수정

- Car의 입장으로 생각하여 메서드와 메서드의 파라미터명 수정

Seongwon97 committed:
refactor(RandomNumber.java): 난수 생성 메서드를 정적으로 사용할 수 있도록 변경

- 메모리, 인스턴스 생성 측면에서의 이점을 생각하여 싱글톤으로 리펙토링 하였었으나
Cars.java와의 강한 결합이 생겨 싱글톤 패턴을 제거하고 정적으로 사용할 수 있게 변경

Seongwon97 committed:
refactor(Input.java, Output.java): 메서드를 정적으로 사용할 수 있도록 변경

- 메서드를 정적으로 사용할 수 있게 변경하여 Controller와  View를 분리

Seongwon97 committed:
refactor(Cars.java, Output.java): view와 연관된 Cars의 메서드 수정

- 출력 String을 내부에서 가공하여 return 하던 Cars의 메서드들을
결과값 list로 결과를 반환하게 수정 후 Output.java에서 결과를 출력하도록 수정
- CarsTest.java파일 삭제

Seongwon97 commented:
안녕하세요 미르 :)

미르의 피드백을 받고 이번에 제출한 2단계 코드가 이전 1단계 제출 코드보다 MVC패턴과 더욱 거리가 멀어졌다는 것을 깨닫고 다시 한번 리펙토링을 해보았습니다.

크게 수정한 부분은 다음과 같습니다.
1. 메서드와 파라미터의 네이밍 변경 (각각의 클래스 기준으로 보도록!!)
2. Input, Output 클래스의 메서드를 정적 메서드로 변경을 하여 Model, View, Controller간의 엮임을 제거
2.1. Model에 위치한 View와 관련된 작업들을 모두 view로 이전
3. RandomNumber 클래스의 메서드를 정적 메서드로 변경하여 Cars객체와의 결합도를 줄임

리뷰를 받은 후로 크게 이러한 작업을 한 것 같습니다.

확실히 리뷰를 반영하니 이전의 코드보다는 Model, View, Controller 간의 관계가 더욱 분리된 것처럼 보이는거 같네요🤓

이번 주말에는 2주동안 학습한 내용 복습, 미르의 피드백 정리 및 크루들의 코드 리뷰를 해볼까합니다.
주말에 학습을 하며 추가적으로 리펙터링을 진행하고 궁금한 내용이 생긴다면 추가로 질문 드리겠습니다😁

감사합니다🙇🏻‍♂️

ddu0422 commented on src/main/java/racingcar/models/Cars.java:
```diff
@@ -0,0 +1,51 @@
+                .filter(car -> car.getPosition() == farthestPosition)
+                .map(car -> car.getName())
+                .collect(Collectors.toList());
+    }
```
해당 부분에 Car 객체에 만들어 둔 메서드를 사용해보면 어떨까요??

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -32,43 +24,22 @@ public void startGame() {
-                .filter((car) -> farthestPosition == car.getPosition())
-                .map(Car::toString)
-                .collect(Collectors.joining(", "));
+        Output.printWinners(cars.findWinners());
```
정적 메서드로 변환한 건 정말 좋습니다 !_! 💯 
사용하는 위치만 조금 변경하면 될 것 같아요!

Input과 Ouput을 Controller가 책임을 지는 게 맞을까요 !_?

[MVC 각 컴포넌트 역할](https://tecoble.techcourse.co.kr/post/2021-04-26-mvc/)과 [MVC 개념](https://pjh3749.tistory.com/27)을 참고해보면 좋을 것 같아요 :)
참고 내용을 보시면서 궁금한 점 있으시면 코멘트가 DM 주세요 ㅎ_ㅎ

ddu0422 reviewed (changes requested):
안녕하세요 렉스 주말에도 공부하신다니 정말 대단하시군요 :)
공부에 도움이 될 수 있게 작은 코멘트 하나 드렸어요 ㅎ_ㅎ
주말이 아니더라도 시간이 아주 조금 있으니 고민해보면 좋을 것 같아요 !_!

궁금한 점은 언제나 코멘트나 DM 주세요!

Seongwon97 committed:
refactor(Cars.java): 우승자를 찾는 메서드의 조건 부분 수정

- 우승자 판별 메서드의 조건을 Car클래스에서 생성한 메서드를 사용하도록 수정

Seongwon97 committed:
refactor(CarDto.java): Cars의 getter메서드를 dto로 반환하도록 수정

- CarDto를 생성하여 Cars.java에서 List<Car>를 반환하던 것을 dto를 통해 반환하도록 수정

Seongwon97 committed:
refactor(Output.java): 불필요한 import제거

Seongwon97 committed:
refactor(src/): 불필요한 import제거, 컨벤션 통일

Seongwon97 commented:
안녕하세요 미르
이번 주말에는 코드 수정보다는 다른 크루들의 코드 리뷰를 중점적으로 공부하여 변경 사항이 적은 것 같습니다.

학습을 하며 많은 내용이 궁금하였으나 DM으로 친절히 알려주셔서 감사합니다.
[983d637](https://github.com/woowacourse/java-racingcar/pull/373/commits/983d6378f653b1dc0e84c768001ad3f644b12d4a) 커밋의 경우는 반환을 DTO로 변경하였으나 추가적인 학습을 해보니 내부값을 보호하기 위해서는 꼭 DTO를 사용하지 않아도 `Collections.unmodifiableList(cars);`를 반환하면 일급 컬렉션의 불변을 보장할 수 있다는 점을 알게 되었어요.
dto의 사용, `Collections.unmodifiableList();`의 사용 모두 객체 내부 값을 보호 할 수 있는 방법이라 생각하여 최종 코드에는 dto를 사용한 코드로 제출을 해볼까 합니다.

1주차 미션기간동안 부족함이 많은 코드를 하나씩 리뷰해주셔서 감사합니다.
덕분에 많은 성장을 할 수 있던것 같습니다.
앞으로 남은 우테코 기간동안도 많은 성장을 하는 개발자가 되도록 하겠습니다.

감사합니다🙇🏻‍♂️

ddu0422 reviewed (approved):
안녕하세요 렉스 :)
🎆 🎉 첫 미션 고생하셨습니다 🎉 🎆

이만 머지해도 될 것 같아서 머지할게요!

고민하셨다가 까먹은 부분도 앞으로 미션하면서 기억나구 더 많은 걸 배울 수 있을거예요 :)

나중에 공부하셔도 되지만 한 가지 말씀드리고 싶은 부분은 `Collections.unmodifiableList(cars);` 가 정말 내부의 객체 보호할 수 있는지 확인해보면 많은 도움이 될 것 같아요 !_!

다음 미션에서도 많은 걸 배울 수 있을거예요 :)
앞으로도 화이팅이예요 ㅎ_ㅎ
